### PR TITLE
GH Action: merge staging(-next) periodically

### PR DESCRIPTION
**Pull request by FRidh**
Automate the merging of `master` -> `staging-next` -> `staging`.

Our main development branch is `master`. Large rebuilds go to `staging`.
Periodically, `staging` is merged into `staging-next` for stabilization.
When considered sufficiently stable, `staging-next` is merged into
`master`.

As changes arrive on these branches, it is important that they're all
updated regularly with eachothers changes. This commit automates that
part.

https://github.com/NixOS/nixpkgs/issues/104594

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SuperSandro2000** [comment]
How does this handle merge conflicts and conflicting/undoing changes? Rebasing is also not an option because it could easily create merge conflicts.

**Comment by FRidh** [comment]
If there is a merge conflict I would *expect* this action to fail, resulting in a notification (but for who?).

This probably should be tested on another repo first.

**Comment by lukegb** [comment]
You probably want to restrict this to run on the main NixOS/nixpkgs repo (e.g. `if: github.repository == ‘NixOS/nixpkgs’`)

**Comment by FRidh** [comment]
Configured it on my fork https://github.com/FRidh/nixpkgs/commit/75fe13617a0b3fa9b03dd230429bc6a21c5b2c68.

File: .github/workflows/merge-staging.yml [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+    # * is a special character in YAML so you have to quote this string
+    # Merge every 6 hours
+    - cron:  '* */6 * * *'
+
+jobs:
+  sync-branch:
+    if: github.repository == 'NixOS/nixpkgs'
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@master
```
Comment (zowoq) [inline .github/workflows/merge-staging.yml:14]:
```suggestion
      - uses: actions/checkout@v2
```
Upstream seems to recommend pinning this in their examples. https://github.com/actions/checkout

**Comment by FRidh** [comment]
Example of a merge conflict https://github.com/FRidh/nixpkgs/runs/1467125543?check_suite_focus=true.

File: .github/workflows/merge-staging.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
+name: "merge staging(-next)"
+
+on:
+  schedule:
+    # * is a special character in YAML so you have to quote this string
+    # Merge every 6 hours
+    - cron:  '* */6 * * *'
```
Comment (Mic92) [inline .github/workflows/merge-staging.yml:7]: How about publishing merge conflicts notifications on IRC?

Comment (FRidh) [reply .github/workflows/merge-staging.yml:7]: That's a good idea. Notifying on IRC could be done with e.g. https://github.com/marketplace/actions/irc-message-action

**Comment by FRidh** [comment]
I added a step that would notify on an issue (currently this one) when the workflow fails.

**Comment by Mic92** [comment]
sounds good.

**Comment by lukegb** [comment]
> I added a step that would notify on an issue (currently this one) when the workflow fails.

Hooray, it worked :^)

**Comment by Mic92** [comment]
> An automatic merge [failed](https://github.com/NixOS/nixpkgs/actions/runs/403256027).

fixed.

**Comment by FRidh** [comment]
> An automatic merge [failed](https://github.com/NixOS/nixpkgs/actions/runs/415367940).

Fixed with a717b55392ffd9c558b6135662c894fa69d4d82d.

**Comment by primeos** [comment]
Any particular reason why the recent automatic merge failures are being ignored?
(Note: Please don't take this as a complaint. I'm just trying to understand our new workflow and wondering if this is intentional (e.g. during staging-next PRs) or if it's due to time constraints, notification issues, etc.).

**Comment by jtojnar** [comment]
Probably no one got around to resolving them, will do that now.

**Comment by jtojnar** [comment]
Done in f19eb635b45.

**Comment by mweinelt** [comment]
How could we reimplement that logic efficiently for release-21.05/staging-21.05/staging-next-21.05?

**Comment by Mic92** [comment]
it would be great if we could convert this to a prometheus metric rather than an issue to see if resolves itself.

**Comment by chrissyjai** [comment]
why am i suscribed to this

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by jtojnar** [comment]
@chrissyjai Maybe https://github.com/NixOS/nixpkgs/issues/165471?

**Comment by mdaniels5757** [comment]
Future failure notifications will happen at #562905; we've reached the limit here. Please consider subscribing.

**Pull request merged**